butterfree/load/writers/writer.py

```diff
@@ -50,3 +70,104 @@ def validate(
             AssertionError: if validation fails.
 
         """
+
+    def write(
+        self, feature_set: FeatureSet, dataframe: DataFrame, spark_client: SparkClient,
+    ) -> Optional[StreamingQuery]:
+        """Loads the latest data from a feature set into the Feature Store.
+
+        Feature Store could be Online or Historical.
+
+        Args:
+            feature_set: object processed with feature set metadata.
+            dataframe: Spark dataframe containing data from a feature set.
+            spark_client: client for Spark connections with external services.
+
+        If the debug_mode is set to True, a temporary table with a name in the format:
+        {feature_set.name} will be created instead of writing
+        to the real historical or online feature store.
+        If dataframe is streaming this temporary table will be updated in real time.
+
+        Returns:
+            Streaming handler if writing streaming df, None otherwise.
+
+        """
+        load_df, db_config, options, database, table_name, partition_by = self.load(
+            feature_set=feature_set, dataframe=dataframe, spark_client=spark_client,
+        )
+
+        pre_hook_df = self.run_pre_hooks(load_df)
+
+        if self.debug_mode:
+            return self._write_in_debug_mode(
+                table_name=(
+                    f"historical_feature_store__{table_name}"
+                    if partition_by
+                    else f"online_feature_store__{table_name}"
+                ),
+                dataframe=pre_hook_df,
+                spark_client=spark_client,
+            )
+
+        if dataframe.isStreaming:
+            return self._write_stream(
+                feature_set=feature_set,
+                dataframe=pre_hook_df,
+                spark_client=spark_client,
+                table_name=table_name,
+                db_config=db_config,
+            )
+
+        spark_client.write_dataframe(
+            dataframe=pre_hook_df,
+            format_=db_config.format_,
+            mode=db_config.mode,
+            **options,
+            partitionBy=partition_by,
+        )
+
+        if partition_by:
+            partition_values = extract_partition_values(pre_hook_df, partition_by)
+
+            spark_client.add_table_partitions(
+                partition_values, feature_set.name, database
+            )
```

Comment:
I didn't covert all the files in this PR, I stopped at this file when I realized we got some misunderstanding from our last talk with @roelschr. Personally I did not understand much of the decisions here, from your last changes and from our talk.

Some problems I'm thinking here:
- I think the speciliazed writers still need to be with the responsabilities to finally connect to the source and write the results, this step is specific for each context. Write partitions, write as stream and other specifications need to be handled by each specific writer, not all sources support that. I think is problematic having these responsabilities in the "father" class, the base Writer.
- The method is running the pre_hooks after some operations on the df (I thought it needed to be the first thing to run)
- This method is referecing Historical and Online feature store writers, but the base `Writer` class should be agnostic to specific writers.

For what I remember the main thing we align to change was not making check schema as a pre_hook, and leaving the `Sink` unchanged. So the main changes would take place on the Online and Historical fs writers themselves. I was not expecting seeing the Writer totaly different like this :thinking: .

Again... maybe just I got confused somehow, so let's talk again while we still need to align the conclusion of this epic. I'm waiting for @roelschr inputs too.